source/AtlasGlyphs.cpp

```diff
@@ -0,0 +1,403 @@
+		else
+		{
+			width += advance[previous * GLYPHS + glyph] + KERN;
+			previous = glyph;
+		}
+	}
+	width += advance[previous * GLYPHS];
+	
+	return width;
+}
+
+
+
+double AtlasGlyphs::LineHeight() const
+{
+	return height;
+}
+
+
+
+double AtlasGlyphs::Space() const
+{
+	return space;
+}
+
+
+
+double AtlasGlyphs::Baseline() const
+{
+	return baseline;
+}
+
+
+
+size_t AtlasGlyphs::FindUnsupported(const string &str, size_t pos) const
+{
+	if(Font::CodePointStart(str, pos) != pos)
+		pos = Font::NextCodePoint(str, pos);
+	for(; pos < str.length(); pos = Font::NextCodePoint(str, pos))
+	{
+		char32_t c = Font::DecodeCodePoint(str, pos);
+		// ASCII
+		if(c >= 32 && c < 32 + GLYPHS - 3 && c != '\'' && c != '"')
+			continue;
+		// Curly quotes
+		if(c == Font::LEFT_SINGLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::RIGHT_SINGLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::LEFT_DOUBLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::RIGHT_DOUBLE_QUOTATION_MARK)
+			continue;
+		// Vertical rectangle
+		if(c == Font::WHITE_VERTICAL_RECTANGLE)
+			continue;
+		return pos;
+	}
+	return str.length();
+}
+
+
+
+int AtlasGlyphs::Glyph(char32_t c)
+{
+	// ASCII
+	if(c >= 32 && c < 32 + GLYPHS - 3 && c != '\'' && c != '"')
+		return c - 32;
+	// Curly quotes
+	if(c == Font::LEFT_SINGLE_QUOTATION_MARK)
+		return 96;
+	if(c == Font::RIGHT_SINGLE_QUOTATION_MARK)
+		return '\'' - 32;
+	if(c == Font::LEFT_DOUBLE_QUOTATION_MARK)
+		return 97;
+	if(c == Font::RIGHT_DOUBLE_QUOTATION_MARK)
+		return '"' - 32;
+	// Vertical rectangle
+	return 95;
+}
+
+
+
+void AtlasGlyphs::LoadTexture()
+{
+	glGenTextures(1, &texture);
+	glBindTexture(GL_TEXTURE_2D, texture);
+	
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+	
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, image.Width(), image.Height(), 0,
+		GL_BGRA, GL_UNSIGNED_BYTE, image.Pixels());
+}
+
+
+
+void AtlasGlyphs::CalculateAdvances()
+{
+	// Get the format and size of the surface.
+	int width = image.Width() / GLYPHS;
+	height = image.Height();
+	unsigned mask = 0xFF000000;
+	unsigned half = 0xC0000000;
+	int pitch = image.Width();
+	
+	// advance[previous * GLYPHS + next] is the x advance for each glyph pair.
+	// There is no advance if the previous value is 0, i.e. we are at the very
+	// start of a string.
+	memset(advance, 0, GLYPHS * sizeof(advance[0]));
+	for(int previous = 1; previous < GLYPHS; ++previous)
+		for(int next = 0; next < GLYPHS; ++next)
+		{
+			int maxD = 0;
+			int glyphWidth = 0;
+			uint32_t *begin = image.Pixels();
+			for(int y = 0; y < height; ++y)
+			{
+				// Find the last non-empty pixel in the previous glyph.
+				uint32_t *pend = begin + previous * width;
+				uint32_t *pit = pend + width;
+				while(pit != pend && (*--pit & mask) < half) {}
```

Comment:
Probably unroll this advancement loop, i.e.
```
while(...)
{
    // Finding the last non-empty pixel in the previous glyph.
}
```

Reply:
Inherited from the original code. I'll leave it as is.